File: SortExamples/MergeSort.js
const generateArray = require('./Global').generateArray;

function mergeSort(arr) {
    console.time('MergeSort');
    let count = 0;
    console.log(main(arr));
    console.timeEnd('MergeSort');
    return count;
    function main(arr) {
        if(arr.length === 1) return arr;

        let mid = Math.floor(arr.length/2);
        let left = arr.slice(0, mid);
        let right = arr.slice(mid);
        return merge(arguments.callee(left), arguments.callee(right));
    }
    function merge(left, right) {
        let il = 0,
            rl = 0,
            result = [];
        while(il < left.length && rl < right.length) {
            count++;
            if(left[il] < right[rl]) {
                result.push(left[il++]);
            }
            else {
                result.push(right[rl++]);
            }
        }
        return result.concat(left.slice(il)).concat(right.slice(rl));
    }
}

console.log(mergeSort(generateArray(10000000)));

File: SortExamples/PythonImplementation/MergeSort.py
from Global import generateList

def mergeSort(l):
    def merge(left, right):
        result = []
        i, j = 0, 0
        ll, rl = len(left), len(right)
        while i < ll and j < rl:
            if left[i] < right[j]:
                result.append(left[i])
                i = i + 1
            else :
                result.append(right[j])
                j = j + 1
        result.extend(left[i:])
        result.extend(right[j:])
        return result
    def main(l):
        if len(l) == 1:
            return l
        mid = int(len(l) / 2)
        left = l[0:mid]
        right = l[mid:]
        return merge(main(left), main(right))
    return main(l)

print(mergeSort(generateList(10000)))